### DISCUSSION TITLE
Improve the intake conversation logic

### DISCUSSION THREAD

**Discussion by solyarisoftware** [daveshap/Medical_Intake#6] [Ideas]
Hi Dave, 
first of all, I want to thank you for your perseverance in your values of sharing your work in so many directions. Specifically, thank you for sharing cognitive architecture learning (ACE, etc.) and this specific project, once again.

Your intake prompt is a masterpiece for its conciseness and power: [Medical_Intake Intake Prompt](https://github.com/daveshap/Medical_Intake/blob/main/system_01_intake.md). 

> I love it because is a perfect example of using LLM as "conversational agents" in the sense that the LLM conduct the conversation without a programmatic programming of the dialog flow, instead you just instructed the LLM to conduct himself the dialogue, and with almost a single sentence! That's great and that's my area of R&D. 

I also enjoyed your chat.py for its supreme simplicity and clarity!

I know your main goal now is to maximize simplicity for the proof of concept; nevertheless, I'm perplexed regarding the logic that guides the end of the conversation between the chatbot and the user. Following your code and prompt, if I understand correctly, you put in the user's hands the decision to conclude the conversation, so the user must enter the word "DONE" to continue. See: [chat.py Line 65](https://github.com/daveshap/Medical_Intake/blob/main/chat.py#L65)

This is a simple effective solution but a bit misleading and not the best in terms of user experience. You can avoid this user decision because the LLM could maybe decide by itself when it has gathered all the necessary information (to be verified, I have to test what I'm writing). So my proposal is to add some logic to the original prompt to let the LLM decide himself to conclude the conversation with the patient.

> By the way, as a personal side project, I experimented with a scenario partially similar: I designed a prompt to instruct a chatbot to trigger (say, on a chat/phone) a conversation with a patient possibly affected by COVID. The chatbot asks the patient everyday for some simple measurements (temperature, oxygen saturation) and inquires about the current health status of the patient. When the chatbot has gathered the attributes temperature, saturation, and health status, it says goodbye to the user, closing the conversation and afterward emits a "signal" to a task controller writing a JSON like: 
`{"temp":38,"saturation":90,"health_status":"the user says that ... blablabla"}`. 

So, going back to the intake conversation, the LLM could emit a JSON signal `{ "end_conversation": true }` to be caught by your chat.py program. In this way you delegate the responsibility to evaluate the information intake to the LLM itself, not the user.

You can do all this maybe just adding a rule in the original prompt with something like:

```markdown
Terminate the conversation with the patient when you have gathered all user feedback about their symptoms; 
In this case, ask the user if anything is missing and bid them farewell. 
Afterward, always write: {"end_conversation":true}
```

Of course this is the base idea and the implementation could depend on the specific LLM model capabilities. E.g. chat completion models are not perfectly good to mix text generation and e.g. JSON. TBV.  

What do you think?

---
 ## Use LiteLLM
By the way, minor note: you could substitute direct OpenAI completion API calls using instead [LiteLLM](https://github.com/BerriAI/litellm) in that way, with a minor modification in your code, you let your applications independent from the specific LLM provider.

**Comment by daveshap** [comment]
Thanks for all the feedback, we're still waiting for a project lead though

**Comment by solyarisoftware** [comment]
I know :-), see: https://github.com/daveshap/Medical_Intake/discussions/5#discussioncomment-7095217

**Comment by daveshap** [comment]
Upon rereading I don't really understand the point of this post. DONE is an incredibly simple approach just for demonstration purposes. Imagine that the user has a "I'm done describing my symptoms" button on the UI.

**Comment by solyarisoftware** [comment]
I completely agree that "DONE" is enough for an initial demonstration, but when considering the overall user experience in a conversational application, there's room for improvement, albeit perhaps minor considering the entire application.

In my understanding, the intake system should primarily function as an advanced assistant capable of engaging in natural language conversations with users. Can you imagine if the user interface involved voice-only interactions via a phone? To me, that seems like a promising direction to explore. The use of a button or a placeholder, in my opinion, is a (conversational) design "anti-pattern".

In conversations with a doctor, it's usually the doctor who guides and assesses whether the patient needs to share more symptoms. Patients typically don't need to explicitly say "DONE" (or the intent to having finished the symptom lists). Instead is the doctor that takes the lead in shaping the conversation. At least, this is the common "social practice". So, I envision a virtual conversational system behaving in a similar manner.

These are just my thoughts on the matter.
giorgio

**Reply by solyarisoftware** [reply]
Hi Dave,
 
see also my recent article: [A Conversational Agent with a Single Prompt?](https://www.linkedin.com/pulse/conversational-agent-single-prompt-giorgio-robino-vrppf/)

Maybe these "conversation design" notes are out your main interests, nevertheless I thank you if you woul read my article and please feel free to comment on linkedin if you will.

Thanks again for your sharings
Giorgio
